### PR TITLE
Add batch flag to remote-test-server 

### PR DESCRIPTION
**Pull request by Ayush1325**
When using this flag, the stdout and stderr are sent in a single batch instead of being streamed. It also used `Command::output` instead of `Command::spawn`. This is useful for targets that might support std but not threading (Eg: UEFI).

Signed-off-by: Ayush Singh <ayushsingh1325@gmail.com>

### CODE REVIEW

File: src/tools/remote-test-server/src/main.rs [hunk 1]
```diff
@@ -277,9 +277,15 @@ fn handle_run(socket: TcpStream, work: &Path, tmp: &Path, lock: &Mutex<()>, conf
     let mut stderr = child.stderr.take().unwrap();
     let socket = Arc::new(Mutex::new(reader.into_inner()));
     let socket2 = socket.clone();
-    let thread = thread::spawn(move || my_copy(&mut stdout, 0, &*socket2));
-    my_copy(&mut stderr, 1, &*socket);
-    thread.join().unwrap();
+
+    if config.sequential {
+        my_copy(&mut stdout, 0, &*socket2);
+        my_copy(&mut stderr, 1, &*socket);
```
Comment (Mark-Simulacrum) [inline src/tools/remote-test-server/src/main.rs:283]: This seems like it won't actually work? The reason we need threads is that my_copy uses blocking I/O -- this will fail (hang!) if the buffers fill up and the process expects to be able to continue writing...

Comment (Ayush1325) [reply src/tools/remote-test-server/src/main.rs:283]: I have used this in https://github.com/rust-lang/rust/pull/100316, so I can say that it does work for UEFI. However, it might just be because of how io writing is implemented in UEFI since there are no threads.

Comment (Ayush1325) [reply src/tools/remote-test-server/src/main.rs:282]: @Mark-Simulacrum So after looking around a bit more in the code, I'm not sure how my implementation is any worse than the current `sequential` implementation. If the sequential option is enabled, the `handle_run` function is launched in the main thread. This means that `my_copy` for stderr runs on the main thread anyway.

My commit just makes `my_copy` for stdout run on the `main` thread` as well. The only target I have tested this on is `UEFI` so maybe you can point to an example where this will lead to UB?

**Comment by Ayush1325** [comment]
@Mark-Simulacrum ping

File: src/tools/remote-test-server/src/main.rs (hunk 1 above)
Comment (Mark-Simulacrum) [reply src/tools/remote-test-server/src/main.rs:282]: This isn't going to lead to UB, it'll just hang -- and only if the process writes to both stdout and stderr in sufficient quantities to fill up the OS pipe.

Running one of the copies on the main thread is fine, but when we run both of them that leads to a problem, because only one of them is running at a time.

One approach to not use threads but solve this is to use a nonblocking API -- you can steal e.g. https://github.com/rust-lang/rust/blob/master/src/tools/compiletest/src/read2.rs to read/write from both on the same thread (always).

File: src/tools/remote-test-server/src/main.rs [hunk 2]
```diff
@@ -336,23 +341,38 @@ fn set_permissions(path: &Path) {
...
-            (n >> 24) as u8,
-            (n >> 16) as u8,
-            (n >> 8) as u8,
-            (n >> 0) as u8,
-        ]));
-        if n > 0 {
-            t!(dst.write_all(&b[..n]));
-        } else {
-            break;
+        match (read_step(out_pipe, 0, dst, &mut b), read_step(err_pipe, 1, dst, &mut b)) {
```
Comment (Mark-Simulacrum) [inline src/tools/remote-test-server/src/main.rs:347]: This isn't enough. You're still blocking on the read from the stdout pipe here, which means that if the process only writes to stderr it will hang. The read2 file I linked handles that by setting the file descriptors to non-blocking, but you're not doing that here.

Comment (Ayush1325) [reply src/tools/remote-test-server/src/main.rs:347]: Well, the read2 implementation only does that for unix and windows. On all the other targets, it uses [this](https://github.com/rust-lang/rust/blob/a161a7b654083a881b22908a475988bcc3221a79/src/tools/compiletest/src/read2.rs#L127)

Comment (Mark-Simulacrum) [reply src/tools/remote-test-server/src/main.rs:347]: Is the remote test server here running in the qemu image (i.e., maybe not on unix or windows)?

In any case, that version of read2 has the same problem as here, but we're rarely running cargo or compiletest outside of cfg(unix) or cfg(windows), maybe even never.

**Comment by pietroalbini** [comment]
My need for sequential execution is, the emulator I'm working with sometimes deadlocks when multiple tests are executed at the same time, but it supports threads just fine (just can't run multiple commands in parallel).

I don't have good ideas on how to implement this without threads, other than maybe adding a `--batch` flag that doesn't stream the output and instead sends stdout and sterr as a whole when the command ends.

r? @Mark-Simulacrum back

**Comment by Ayush1325** [comment]
@Mark-Simulacrum How about adding the `--batch` flag like @pietroalbini  suggested and using `Command::output` instead of `Command::spawn` in it? I think that could be useful after https://github.com/rust-lang/rust/pull/105458.

**Comment by Mark-Simulacrum** [comment]
That sounds good to me.

File: src/tools/remote-test-server/src/main.rs [hunk 3]
```diff
@@ -356,6 +368,17 @@ fn my_copy(src: &mut dyn Read, which: u8, dst: &Mutex<dyn Write>) {
     }
 }
 
+fn batch_copy(buf: &[u8], which: u8, dst: &Mutex<dyn Write>) {
+    let n = buf.len();
+    let mut dst = dst.lock().unwrap();
+    t!(dst.write_all(&[which, (n >> 24) as u8, (n >> 16) as u8, (n >> 8) as u8, (n >> 0) as u8,]));
```
Comment (Mark-Simulacrum) [inline src/tools/remote-test-server/src/main.rs:374]: As a drive-by improvement since we're changing this code, can we replace the manual shifting with writing `(n as u32).encode_le_bytes()` and `read_u32` to use `u32::from_le_bytes`?

Comment (Mark-Simulacrum) [reply src/tools/remote-test-server/src/main.rs:374]: (But let's do that in a separate PR if you're up to it).
